src/sinks/pulsar/service.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use pulsar::producer::Message;
+use pulsar::{Error as PulsarError, Executor, MultiTopicProducer, ProducerOptions, Pulsar};
+use tokio::sync::Mutex;
+use tower::Service;
+use vector_common::internal_event::CountByteSize;
+use vector_core::stream::DriverResponse;
+
+use crate::event::{EventFinalizers, EventStatus, Finalizable};
+use crate::internal_events::PulsarSendingError;
+use crate::sinks::pulsar::request_builder::PulsarMetadata;
+use vector_common::request_metadata::{MetaDescriptive, RequestMetadata};
+
+#[derive(Clone)]
+pub(super) struct PulsarRequest {
+    pub body: Bytes,
+    pub metadata: PulsarMetadata,
+    pub request_metadata: RequestMetadata,
+}
+
+pub struct PulsarResponse {
+    event_byte_size: usize,
+}
+
+impl DriverResponse for PulsarResponse {
+    fn event_status(&self) -> EventStatus {
+        EventStatus::Delivered
+    }
+
+    fn events_sent(&self) -> CountByteSize {
+        CountByteSize(1, self.event_byte_size)
+    }
+
+    fn bytes_sent(&self) -> Option<usize> {
+        Some(self.event_byte_size)
+    }
+}
+
+impl Finalizable for PulsarRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.metadata.finalizers)
+    }
+}
+
+impl MetaDescriptive for PulsarRequest {
+    fn get_metadata(&self) -> RequestMetadata {
+        self.request_metadata
+    }
+}
+
+#[derive(Clone)]
+pub struct PulsarService<Exe: Executor> {
+    producer: Arc<Mutex<MultiTopicProducer<Exe>>>,
+}
+
+impl<Exe: Executor> PulsarService<Exe> {
+    pub(crate) fn new(
+        pulsar_client: Pulsar<Exe>,
+        producer_options: ProducerOptions,
+        producer_name: Option<String>,
+    ) -> PulsarService<Exe> {
+        let mut builder = pulsar_client.producer().with_options(producer_options);
+
+        if let Some(name) = producer_name {
+            builder = builder.with_name(name);
+        }
+
+        let producer = builder.build_multi_topic();
+
+        PulsarService {
+            producer: Arc::new(Mutex::new(producer)),
+        }
+    }
+}
+
+impl<Exe: Executor> Service<PulsarRequest> for PulsarService<Exe> {
+    type Response = PulsarResponse;
+    type Error = PulsarError;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
```

Comment:
Something like this should make things behave better with the mutex:

```suggestion
        if self.producer.try_lock().is_ok() {
            Poll::Ready(Ok(()))
        } else {
            Poll::Pending
        }
```